tests/devices/qutrit_mixed/test_qutrit_mixed_preprocessing.py

```diff
@@ -0,0 +1,278 @@
+
+    def test_chooses_best_gradient_method(self):
+        """Test that preprocessing chooses backprop as the best gradient method."""
+        dev = DefaultQutritMixed()
+
+        config = ExecutionConfig(
+            gradient_method="best", use_device_gradient=None, grad_on_execution=None
+        )
+
+        _, new_config = dev.preprocess(config)
+
+        assert new_config.gradient_method == "backprop"
+        assert new_config.use_device_gradient
+        assert not new_config.grad_on_execution
+
+    def test_circuit_wire_validation(self):
+        """Test that preprocessing validates wires on the circuits being executed."""
+        dev = DefaultQutritMixed(wires=3)
+
+        circuit_valid_0 = qml.tape.QuantumScript([qml.TShift(0)])
+        program, _ = dev.preprocess()
+        circuits, _ = program([circuit_valid_0])
+        assert circuits[0].circuit == circuit_valid_0.circuit
+
+        circuit_valid_1 = qml.tape.QuantumScript([qml.TShift(1)])
+        program, _ = dev.preprocess()
+        circuits, _ = program([circuit_valid_0, circuit_valid_1])
+        assert circuits[0].circuit == circuit_valid_0.circuit
+        assert circuits[1].circuit == circuit_valid_1.circuit
+
+        invalid_circuit = qml.tape.QuantumScript([qml.TShift(4)])
+        with pytest.raises(qml.wires.WireError, match=r"Cannot run circuit\(s\) on"):
+            program, _ = dev.preprocess()
+            program(
+                [
+                    invalid_circuit,
+                ]
+            )
+
+        with pytest.raises(qml.wires.WireError, match=r"Cannot run circuit\(s\) on"):
+            program, _ = dev.preprocess()
+            program([circuit_valid_0, invalid_circuit])
+
+    @pytest.mark.parametrize(
+        "mp_fn,mp_cls,shots",
+        [
+            (qml.sample, qml.measurements.SampleMP, 10),
+            (qml.state, qml.measurements.StateMP, None),
+            (qml.probs, qml.measurements.ProbabilityMP, None),
+        ],
+    )
+    def test_measurement_is_swapped_out(self, mp_fn, mp_cls, shots):
+        """Test that preprocessing swaps out any MP with no wires or obs"""
+        dev = DefaultQutritMixed(wires=3)
+        original_mp = mp_fn()
+        exp_z = qml.expval(qml.GellMann(0, 3))
+        qs = qml.tape.QuantumScript([qml.THadamard(0)], [original_mp, exp_z], shots=shots)
+        program, _ = dev.preprocess()
+        tapes, _ = program([qs])
+        assert len(tapes) == 1
+        tape = tapes[0]
+        assert tape.operations == qs.operations
+        assert tape.measurements != qs.measurements
+        assert qml.equal(tape.measurements[0], mp_cls(wires=[0, 1, 2]))
+        assert tape.measurements[1] is exp_z
+
+    @pytest.mark.parametrize(
+        "op, expected",
+        [
+            (qml.TShift(0), True),
+            (qml.GellMann(0, 1), False),
+            (qml.Snapshot(), True),
+            (qml.TRX(1.1, 0), True),
+        ],
+    )
+    def test_accepted_operator(self, op, expected):
```

Comment:
Is it worth leaving a TODO here to add a Channel to this list, once those are ready to go?

Reply:
Yes, I can add that.